src/wp-admin/includes/image-edit.php

```diff
@@ -1026,12 +1061,177 @@ function wp_save_image( $post_id ) {
 		$meta['sizes'] = array_merge( $meta['sizes'], $img->multi_resize( $_sizes ) );
 	}
 
+	$resize_sizes = array();
+	foreach ( $old_metadata['sizes'] as $size_name => $size_details ) {
+		// If the target is 'nothumb', skip generating the 'thumbnail' size.
+		if ( 'nothumb' === $target && 'thumbnail' === $size_name ) {
+			continue;
+		}
+
+		if (
+			isset( $meta['sizes'][ $size_name ] ) && ! empty( $meta['sizes'][ $size_name ] ) &&
+			$meta['sizes'][ $size_name ]['file'] !== $old_metadata['sizes'][ $size_name ]['file']
+		) {
+			$resize_sizes[ $size_name ] = $meta['sizes'][ $size_name ];
+		}
+	}
+
+	$uploads_dir        = wp_get_upload_dir();
+	$allowed_mimes      = array_flip( wp_get_mime_types() );
+	$original_directory = pathinfo( $new_path, PATHINFO_DIRNAME );
+	$filename           = pathinfo( $new_path, PATHINFO_FILENAME );
+	$main_images        = array();
+	$subsized_images    = array();
+
+	$transforms = wp_upload_image_mime_transforms( $post_id, false );
+
+	if ( ! empty( $transforms[ $post->post_mime_type ] ) ) {
+		$mime_transforms = $transforms[ $post->post_mime_type ];
+
+		foreach ( $mime_transforms as $targeted_mime ) {
+			if ( $targeted_mime === $post->post_mime_type ) {
+				// If the target is `thumbnail` make sure it is the only selected size.
+				if ( 'thumbnail' === $target ) {
+					if ( isset( $meta['sizes']['thumbnail'] ) ) {
+						$subsized_images[ $targeted_mime ] = array( 'thumbnail' => $meta['sizes']['thumbnail'] );
+					}
+					// When the targeted thumbnail is selected no additional size and subsize is set.
+					continue;
+				}
+
+				$main_images[ $targeted_mime ]     = array(
+					'path' => $new_path,
+					'file' => pathinfo( $new_path, PATHINFO_BASENAME ),
+				);
+				$subsized_images[ $targeted_mime ] = $meta['sizes'];
+				continue;
+			}
+
+			if ( ! isset( $allowed_mimes[ $targeted_mime ] ) || ! is_string( $allowed_mimes[ $targeted_mime ] ) ) {
+				continue;
+			}
+
+			if ( ! $img::supports_mime_type( $targeted_mime ) ) {
+				continue;
+			}
+
+			$extension = explode( '|', $allowed_mimes[ $targeted_mime ] );
+			$extension = $extension[0];
+
+			// If the target is `thumbnail` make sure only that size is generated.
+			if ( 'thumbnail' === $target ) {
+				if ( ! isset( $subsized_images[ $post->post_mime_type ]['thumbnail']['file'] ) ) {
+					continue;
+				}
+				$thumbnail_file    = $subsized_images[ $post->post_mime_type ]['thumbnail']['file'];
+				$current_extension = pathinfo( $thumbnail_file, PATHINFO_EXTENSION );
+
+				// Create a file with then new extension out of the targeted file.
+				$target_file_name     = preg_replace( "/\.$current_extension$/", ".$extension", $thumbnail_file );
+				$target_file_location = path_join( $original_directory, $target_file_name );
+				$result               = $img->save( $target_file_location, $targeted_mime );
```

Comment:
can you add inline docs indicating which file is being created in each save spot? 

I don't think you need to construct the filename, you can use `set_output_mime_type` and the image editor class will generate the correct filename.

Reply:
this particular block of code, generates/saves thumbnail images as it is in the condition that checks if `'thumbnail' === $target`. In this we can't use just `set_output_mime_type` as it will use image file of full size instead of thumbnail and will generate that file only.